### ISSUE TITLE
Create Macr-to-Macr FSTs

### ISSUE THREAD

**Issue by eddieantonio** [UAlbertaALTLab/itwewina#41]
Just like the Cans-to-Cans FSTs, we need Macr-to-Macr FSTs to workaround the fact that NDS cannot really share dictionaries without EXTENSIVE refactoring, and that the "internal representation" of the FSTs is in circumflexes.

Closing this issue will resolve #38.

**Comment by aarppe** [comment]
1. Analyser that will spout out lemmas according to SRO/macron (but will accept as input SRO with/without length indication or with macrons or circumflexes:

Composition:

    hfst-compose -F -1 src/analyser-gt-desc.hfst -2 src/orthography/circumfl-to-macron.compose.hfst -o analyser-gt-desc.macron.hfst

Examples:
```
hfst-lookup -q analyser-gt-desc.macron.hfst
nipihk
nipihk	nipiy+N+I+Loc	0.000000
nipihk	nīpiy+N+I+Loc	0.000000
nipihk	nipiw+V+AI+Cnj+Prs+X	0.000000
nipihk	nīpin+V+II+Cnj+Prs+3Sg	0.000000
nipihk	nipiw+V+AI+Imp+Imm+2Pl	0.000000

nīpīhk
nīpīhk	nīpiy+N+I+Loc	0.000000

nīpihk
nīpihk	nīpiy+N+I+Loc	0.000000
nīpihk	nīpin+V+II+Cnj+Prs+3Sg	0.000000

nipīhk       
nipīhk	nipiy+N+I+Loc	0.000000
nipīhk	nīpiy+N+I+Loc	0.000000
```

2. Normative generator of SRO macron word forms that expects lemmas with normative SRO macrons for long vowels.

Composition:

    hfst-compose -F -1 macron-to-circumfl.compose.hfst -2 src/generator-gt-norm.hfst -o - | hfst-compose -F -1 - -2 src/orthography/circumfl-to-macron.compose.hfst -o generator-gt-norm.macron-to-macron.hfst

Examples:
```
hfst-lookup -q generator-gt-norm.macron-to-macron.hfst
nipiy+N+I+Loc
nipiy+N+I+Loc	nipīhk	0.000000

nīpiy+N+I+Loc
nīpiy+N+I+Loc	nīpīhk	0.000000

nipiw+V+AI+Cnj+Prs+X
nipiw+V+AI+Cnj+Prs+X	nipihk	0.000000
nipiw+V+AI+Cnj+Prs+X	nipiniwahk	0.000000
nipiw+V+AI+Cnj+Prs+X	nipināniwahk	0.000000

nīpin+V+II+Cnj+Prs+3Sg
nīpin+V+II+Cnj+Prs+3Sg	nīpihk	0.000000

nipiw+V+AI+Imp+Imm+2Pl
nipiw+V+AI+Imp+Imm+2Pl	nipik	0.000000


```

**Comment by aarppe** [comment]
As noted in #43, implementing orthographical omnivorosity requires one additional composition in the very beginning:

COMPOSITION:

    hfst-compose -F -1 src/orthography/Cans-to-Latn.compose.hfst -2 src/analyser-gt-desc.hfst | hfst-compose -F -1 - -2 src/orthography/circumfl-to-macron.compose.hfst -o analyser-gt-desc.macron.hfst

EXAMPLES:
```
hfst-lookup -q analyser-gt-desc.macron.hfst 
ᓂᐱᕽ
ᓂᐱᕽ	nipiy+N+I+Loc	0.000000
ᓂᐱᕽ	nīpiy+N+I+Loc	0.000000
ᓂᐱᕽ	nipiw+V+AI+Cnj+Prs+X	0.000000
ᓂᐱᕽ	nīpin+V+II+Cnj+Prs+3Sg	0.000000
ᓂᐱᕽ	nipiw+V+AI+Imp+Imm+2Pl	0.000000

niᐱᕽ
niᐱᕽ	nipiy+N+I+Loc	0.000000
niᐱᕽ	nīpiy+N+I+Loc	0.000000
niᐱᕽ	nipiw+V+AI+Cnj+Prs+X	0.000000
niᐱᕽ	nīpin+V+II+Cnj+Prs+3Sg	0.000000
niᐱᕽ	nipiw+V+AI+Imp+Imm+2Pl	0.000000

nipihk
nipihk	nipiy+N+I+Loc	0.000000
nipihk	nīpiy+N+I+Loc	0.000000
nipihk	nipiw+V+AI+Cnj+Prs+X	0.000000
nipihk	nīpin+V+II+Cnj+Prs+3Sg	0.000000
nipihk	nipiw+V+AI+Imp+Imm+2Pl	0.000000

nīpīhk
nīpīhk	nīpiy+N+I+Loc	0.000000

nīpîhk
nīpîhk	nīpiy+N+I+Loc	0.000000
nīpîhk	nīpin+V+II+Cnj+Prs+3Sg	0.000000
```

**Comment by eddieantonio** [comment]
@aarppe I can't seem to find `macron-to-circumfl.compose.hfst`. Where is it? Alternatively, how do I generate it?

**Comment by aarppe** [comment]
Here (under giella/langs/crk/):

src/orthography/circumfl-to-macron.compose.hfst

**Comment by eddieantonio** [comment]
This file does not appear to be committed to the SVN repository:

```
 $ pwd
/Users/eddieantonio/Projects/giella/langs/crk
 $ svn up
Updating '.':
At revision 170693.
 $ svn info src/orthography/circumfl-to-macron.compose.hfst
svn: warning: W155010: The node '/Users/eddieantonio/Projects/giella/langs/crk/src/orthography/circumfl-to-macron.compose.hfst' was not found.

svn: E200009: Could not display info for all targets because some targets don't exist
```

**Comment by aarppe** [comment]
As far as I understand, that FST should not be committed but created when running make in giella/langs/crk

**Comment by eddieantonio** [comment]
Not happening for me for some strange reason. I'll create a bodge rule in src/Makefile.am to generate it. I assume an `hfst-invert` on `circumfl-to-macron.compose.hfst` will work.

**Comment by aarppe** [comment]
You would expect that inversing the FST would work - but you may recall that looking up in the inverse direction produced all the interim results - though maybe composing will do away with that excess.

I am also just checking whether I can compile that FST from scratch from src/orthography/circumfl-to-macron.regex - do you see that file?

An alternative is creating the source regex explicitly in the opposite direction and the compiling that regex.

**Comment by eddieantonio** [comment]
> I am also just checking whether I can compile that FST from scratch from src/orthography/circumfl-to-macron.regex - do you see that file?

Yes! However, I don't see the "inverse" file:  "macron-to-circumfl.regex", which is probably why I don't see its .compose and .lookup variants.

**Comment by aarppe** [comment]
Ok. Appears this was my bad. What had happened was that I had indeed inverted the existing circumfl-to-macron.compose.hfst in src/orthography/, cf.

hfst-invert -i src/orthography/circumfl-to-macron.compose.hfst -o macron-to-circumfl.compose.hfst

... and forgotten that it was not automatically compiled, and then run the composition sequence for creating the macron-to-macron generator (above) without realizing that one part would be missing for everyone else but me. Anyhow, you should put the inverted FST under src/orthography/ for the sake of consistency.

Anyhow, this shows that the inversion for the transcriptor works at least for the purposes of composition, even though the lookup produces weird'ish (for all but FST experts) results in one direction.

**Comment by eddieantonio** [comment]
Fixed as of 01cd626. See #43.
